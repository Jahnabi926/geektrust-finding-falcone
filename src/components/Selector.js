import {
  DestinationDropdown,
  VehicleDropdown,
} from "../helpers/SelectorUtilityFunctions";
import styles from "../styles/selector.module.css";

export default function Selector(props) {
  const {
    planetOptions,
    planetDropdowns,
    vehicleDropdowns,
    handlePlanetToggle,
    handlePlanetSelection,
    handleVehicleSelection,
  } = props;

  return (
    <>
      <div className={styles.container}>
        {planetDropdowns.map((dropdown, index) => (
          <div key={`planet-dropdown-${index}`}>
            <DestinationDropdown
              dropdown={dropdown}
              index={index}
              planetOptions={planetOptions}
              planetDropdowns={planetDropdowns}
              handlePlanetToggle={handlePlanetToggle}
              handlePlanetSelection={handlePlanetSelection}
            />
            <VehicleDropdown
              dropdown={vehicleDropdowns[index]}
              index={index}
              handleVehicleSelection={handleVehicleSelection}
            />
          </div>
        ))}
      </div>
    </>
  );
}
