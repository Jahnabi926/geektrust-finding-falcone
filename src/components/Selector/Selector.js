import { useState, useRef, useEffect } from "react";
import PlanetDropdown from "../PlanetDropdown/PlanetDropdown";
import VehicleDropdown from "../VehicleDropdown/VehicleDropdown";
import styles from "../Selector/selector.module.css";
import React from "react";

export default function Selector(props) {
  const {
    planetOptions,
    planetDropdowns,
    vehicleDropdowns,
    handlePlanetToggle,
    handlePlanetSelection,
    handleVehicleSelection,
  } = props;

  const [userInputs, setUserInputs] = useState(planetDropdowns.map(() => ""));

  // Function to handle user input change for a specific planet dropdown
  const handleInput = (index, value) => {
    const newInputs = [...userInputs];
    newInputs[index] = value;
    setUserInputs(newInputs);
  };

  // Handle click outside
  const selectorRef = useRef(null);

  useEffect(() => {
    const handleClickOutside = (event) => {
      if (selectorRef.current && !selectorRef.current.contains(event.target)) {
        // Logic to close all open dropdowns when clicking outside
        planetDropdowns.forEach((dropdown, index) => {
          if (dropdown.isOpen) {
            handlePlanetToggle(index);
          }
        });
      }
    };

    document.addEventListener("mousedown", handleClickOutside);

    return () => {
      document.removeEventListener("mousedown", handleClickOutside);
    };
  }, [planetDropdowns, handlePlanetToggle]);

  return (
    <div className={styles.selector}>
      <div className={styles.outer_wrap} ref={selectorRef}>
        {planetDropdowns.map((dropdown, index) => (
          <div key={`planet-dropdown-${index}`} className={styles.wrap}>
            <PlanetDropdown
              dropdown={dropdown}
              index={index}
              planetOptions={planetOptions}
              planetDropdowns={planetDropdowns}
              handlePlanetToggle={handlePlanetToggle}
              handlePlanetSelection={handlePlanetSelection}
              userInput={userInputs[index]}
              handleInput={(value) => handleInput(index, value)}
            />
            <VehicleDropdown
              dropdown={vehicleDropdowns[index]}
              index={index}
              handleVehicleSelection={handleVehicleSelection}
            />
          </div>
        ))}
      </div>
    </div>
  );
}
