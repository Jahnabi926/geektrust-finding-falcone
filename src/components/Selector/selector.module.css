.selector {
  display: flex;
  justify-content: center;
}
.outer_wrap {
  display: flex;
  gap: 20px;
}
