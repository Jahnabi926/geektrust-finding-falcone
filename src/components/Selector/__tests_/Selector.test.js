import { fireEvent, render, screen } from "@testing-library/react";
import Selector from "../Selector";
import renderer from "react-test-renderer";

const mockPlanetOptions = [
  {
    name: "Donlon",
    distance: 100,
  },
  {
    name: "Enchai",
    distance: 200,
  },
  {
    name: "Jebing",
    distance: 300,
  },
  {
    name: "Sapir",
    distance: 400,
  },
  {
    name: "Lerbin",
    distance: 500,
  },
  {
    name: "Pingasor",
    distance: 600,
  },
];

const mockVehicleOptions = [
  {
    name: "Space pod",
    total_no: 2,
    max_distance: 200,
    speed: 2,
  },
  {
    name: "Space rocket",
    total_no: 1,
    max_distance: 300,
    speed: 4,
  },
  {
    name: "Space shuttle",
    total_no: 1,
    max_distance: 400,
    speed: 5,
  },
  {
    name: "Space ship",
    total_no: 2,
    max_distance: 600,
    speed: 10,
  },
];

const mockPlanetDropdowns = [
  {
    isOpen: false,
    selectedPlanet: null,
    id: "d1",
    timeTaken: 0,
    filteredPlanetOptions: [],
  },
  {
    isOpen: false,
    selectedPlanet: null,
    id: "d2",
    timeTaken: 0,
    filteredPlanetOptions: [],
  },
  {
    isOpen: false,
    selectedPlanet: null,
    id: "d3",
    timeTaken: 0,
    filteredPlanetOptions: [],
  },
  {
    isOpen: false,
    selectedPlanet: null,
    id: "d4",
    timeTaken: 0,
    filteredPlanetOptions: [],
  },
];

const mockVehicleDropdowns = [
  {
    isOpen: false,
    currentSelectedVehicle: null,
    previousSelectedVehicle: null,
    id: "v1",
    associatedPlanetDropdown: "d1",
    selectedPlanet: null,
  },
  {
    isOpen: false,
    currentSelectedVehicle: null,
    previousSelectedVehicle: null,
    id: "v2",
    associatedPlanetDropdown: "d2",
    selectedPlanet: null,
  },
  {
    isOpen: false,
    currentSelectedVehicle: null,
    previousSelectedVehicle: null,
    id: "v3",
    associatedPlanetDropdown: "d3",
    selectedPlanet: null,
  },
  {
    isOpen: false,
    currentSelectedVehicle: null,
    previousSelectedVehicle: null,
    id: "v4",
    associatedPlanetDropdown: "d4",
    selectedPlanet: null,
  },
];

const mockHandlePlanetToggle = jest.fn();
const mockHandlePlanetSelection = jest.fn();
const mockHandleVehicleSelection = jest.fn();

const mockProps = {
  planetOptions: mockPlanetOptions,
  planetDropdowns: mockPlanetDropdowns,
  setPlanetDropdowns: jest.fn(),
  vehicleDropdowns: mockVehicleDropdowns,
  vehicleOptions: mockVehicleOptions,
  handlePlanetToggle: mockHandlePlanetToggle,
  handlePlanetSelection: mockHandlePlanetSelection,
  handleVehicleSelection: mockHandleVehicleSelection,
};

describe("Selector Component", () => {
  test("renders Selector without errors", () => {
    render(<Selector {...mockProps} />);
  });

  test("displays planet and vehicle dropdowns correctly", () => {
    render(<Selector {...mockProps} />);

    //Asserts that the rendered planet dropdowns are displayed correctly
    mockPlanetDropdowns.forEach((_, index) => {
      expect(screen.getByText(`Destination ${index + 1}`)).toBeInTheDocument();
    });
  });

  test("closes all open dropdowns when clicking outside", () => {
    const { container } = render(
      // Wraps the component with a div that is outside of the Selector component
      <div>
        <Selector {...mockProps} />
      </div>
    );

    // Simulates a click outside by clicking on the container div
    fireEvent.mouseDown(container);

    //Asserts that handlePlanetToggle has been called for each open dropdown
    setTimeout(() => {
      mockPlanetDropdowns.forEach((_, index) => {
        expect(mockHandlePlanetToggle).toHaveBeenCalledWith(
          index,
          mockPlanetDropdowns,
          mockProps.setPlanetDropdowns
        );
      });
    }, 0);
  });

  test("matches snapshot", () => {
    const component = renderer.create(<Selector {...mockProps} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
