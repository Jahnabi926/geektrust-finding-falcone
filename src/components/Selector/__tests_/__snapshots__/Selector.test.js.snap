// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Selector Component matches snapshot 1`] = `
<div
  className="selector"
>
  <div
    className="outer_wrap"
  >
    <div
      className="wrap"
    >
      <p>
        Destination 
        1
      </p>
      <button
        className="dropdown_button"
        onClick={[Function]}
      >
        Select an option
        <img
          alt="Caret Icon"
          className="caret"
          preload="auto"
          src="caret-icon.jpg"
        />
      </button>
    </div>
    <div
      className="wrap"
    >
      <p>
        Destination 
        2
      </p>
      <button
        className="dropdown_button"
        onClick={[Function]}
      >
        Select an option
        <img
          alt="Caret Icon"
          className="caret"
          preload="auto"
          src="caret-icon.jpg"
        />
      </button>
    </div>
    <div
      className="wrap"
    >
      <p>
        Destination 
        3
      </p>
      <button
        className="dropdown_button"
        onClick={[Function]}
      >
        Select an option
        <img
          alt="Caret Icon"
          className="caret"
          preload="auto"
          src="caret-icon.jpg"
        />
      </button>
    </div>
    <div
      className="wrap"
    >
      <p>
        Destination 
        4
      </p>
      <button
        className="dropdown_button"
        onClick={[Function]}
      >
        Select an option
        <img
          alt="Caret Icon"
          className="caret"
          preload="auto"
          src="caret-icon.jpg"
        />
      </button>
    </div>
  </div>
</div>
`;
